cdc/owner/scheduler.go

```diff
@@ -0,0 +1,301 @@
+		zap.Int("tableID", int(tableID)),
+		zap.Int("ttl", ttl))
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	s.affinities[tableID] = &affinity{
+		targetCapture: captureID,
+		deadline:      time.Now().Add(time.Duration(ttl) * time.Second),
+	}
+}
+
+func (s *schedulerImpl) onNewCapture(_ model.CaptureID) {
+	s.needRebalance = true
+}
+
+func (s *schedulerImpl) lookUpAffinity(tableID model.TableID) model.CaptureID {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	s.cleanUpAffinities()
+
+	af, ok := s.affinities[tableID]
+	if !ok {
+		return ""
+	}
+
+	if !s.ownerState.CaptureExists(af.targetCapture) {
+		delete(s.affinities, tableID)
+		return ""
+	}
+
+	return af.targetCapture
+}
+
+// cleanUpAffinities must be called with mu locked.
+func (s *schedulerImpl) cleanUpAffinities() {
+	for tableID, af := range s.affinities {
+		if af.deadline.Before(time.Now()) {
+			delete(s.affinities, tableID)
+		}
+	}
+}
+
+func (s *schedulerImpl) triggerRebalance() {
+	tableToCaptureMap := s.ownerState.GetTableToCaptureMap(s.cfID)
+	totalTableNum := len(tableToCaptureMap)
+	captureNum := len(s.ownerState.Captures)
+
+	upperLimitPerCapture := int(math.Ceil(float64(totalTableNum) / float64(captureNum)))
+
+	log.Info("Start rebalancing",
+		zap.String("cfID", s.cfID),
+		zap.Int("table-num", totalTableNum),
+		zap.Int("capture-num", captureNum),
+		zap.Int("target-limit", upperLimitPerCapture))
+
+	for captureID := range s.ownerState.Captures {
+		captureTables := s.ownerState.GetCaptureTables(s.cfID, captureID)
+
+		// Use rand.Perm as a randomization source for choosing victims uniformly.
+		randPerm := rand.Perm(len(captureTables))
+
+		for i := 0; i < len(captureTables)-upperLimitPerCapture; i++ {
+			victimIdx := randPerm[i]
+			victimTableID := captureTables[victimIdx]
+
+			log.Info("triggerRebalance: Stopping table",
+				zap.Int64("table-id", victimTableID),
+				zap.String("capture", captureID),
+				zap.String("changefeed-id", s.cfID))
+
+			s.ownerState.StartDeletingTable(s.cfID, captureID, victimTableID)
+		}
+	}
+}
+
+func (s *schedulerImpl) getMinWorkloadCapture() model.CaptureID {
+	workloads := make(map[model.CaptureID]int)
+
+	for captureID := range s.ownerState.Captures {
+		workloads[captureID] = s.captureWorkloadDeltas[captureID]
+	}
+
+	for _, captureStatuses := range s.ownerState.TaskStatuses {
+		for captureID, taskStatus := range captureStatuses {
+			if _, ok := workloads[captureID]; ok {
+				workloads[captureID] += len(taskStatus.Tables)
+			}
+		}
+	}
+
+	minCapture := ""
+	minWorkLoad := math.MaxInt32
+	for captureID, workload := range workloads {
+		if workload < 0 {
+			// TODO investigate and remove this log
+			log.Debug("negative workload, bug?",
+				zap.Reflect("workloads", workloads),
+				zap.Reflect("deltas", s.captureWorkloadDeltas))
+			workload = 0
+		}
+
+		if workload < minWorkLoad {
+			minCapture = captureID
+			minWorkLoad = workload
+		}
+	}
+
+	return minCapture
+}
+
+func (s *schedulerImpl) cleanUpStaleCaptureStatus() bool {
```

Comment:
please add comment about the meaning of return value, since it is not straightforward.